src/components/app/ServiceWorkerManager.js

```diff
@@ -0,0 +1,152 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+import React, { PureComponent } from 'react';
+import explicitConnect from '../../utils/connect';
+
+import { getDataSource } from '../../reducers/url-state';
+import { getView } from '../../reducers/app';
+
+import type {
+  ExplicitConnectOptions,
+  ConnectedProps,
+} from '../../utils/connect';
+import type { DataSource } from '../../types/actions';
+import type { Phase } from '../../types/reducers';
+
+import './ServiceWorkerManager.css';
+
+type StateProps = {|
+  +dataSource: DataSource,
+  +phase: Phase,
+|};
+type Props = ConnectedProps<{||}, StateProps, {||}>;
+
+type InstallStatus = 'pending' | 'ready' | 'idle';
+type State = {|
+  installStatus: InstallStatus,
+  closeNotice: boolean,
+|};
+
+class ServiceWorkerManager extends PureComponent<Props, State> {
+  state = {
+    installStatus: 'idle',
+    closeNotice: true,
+  };
+
+  _installServiceWorker() {
+    // We use '@mstange/offline-plugin/runtime' here instead of
+    // 'offline-plugin/runtime' because the fork contains the fix from
+    // https://github.com/NekR/offline-plugin/pull/410
+    const runtime = require('@mstange/offline-plugin/runtime');
+    runtime.install({
+      onInstalled: () => {
+        console.log('[ServiceWorker] App is ready for offline usage!');
+      },
+      onUpdating: () => {
+        console.log(
+          '[ServiceWorker] An update has been found and the browser is downloading the new assets.'
+        );
+      },
+      onUpdateReady: () => {
+        console.log(
+          '[ServiceWorker] We have downloaded the new assets and we are ready to go.'
+        );
+        runtime.applyUpdate();
+        this.setState({ installStatus: 'pending' });
+      },
+      onUpdated: () => {
+        console.log(
+          '[ServiceWorker] The new version of the application has been enabled.'
+        );
+        this.setState({ installStatus: 'ready', closeNotice: false });
+      },
+      onUpdateFailed: () => {
+        console.log(
+          '[ServiceWorker] We failed to update the application for an unknown reason.'
+        );
+      },
+    });
+  }
+
+  componentWillMount() {
+    if (process.env.NODE_ENV === 'production') {
+      this._installServiceWorker();
+    }
+  }
+
+  componentDidUpdate() {
+    const { phase } = this.props;
+    const { installStatus } = this.state;
+
+    if (phase !== 'FATAL_ERROR') {
+      return;
+    }
+
+    if (installStatus === 'ready') {
+      this.reloadPage();
+    }
+  }
+
+  reloadPage() {
+    window.location.reload();
+  }
+
+  _onCloseNotice = () => {
+    this.setState({ closeNotice: true });
+  };
+
+  render() {
+    const { dataSource } = this.props;
+    const { installStatus, closeNotice } = this.state;
+
+    if (dataSource !== 'none' && dataSource !== 'public') {
```

Comment:
Hm, I'm not sure what's the difference between `public` and `from-url` as `dataSource` but it seems like we should also add it here.

Also, I tried to reload the page with `from-addon` as well. And apparently we also don't lose any data. It reloads the previous profile successfully.

Reply:
> I'm not sure what's the difference between public and from-url as dataSource

`public` is for profiles that come from the public store. `from-url` is for profiles where we supply a specific URL (we don't have a UI for this yet, this is being worked on in https://github.com/devtools-html/perf.html/pull/1411 by a contributor. There's also some information for this in https://github.com/devtools-html/perf.html/blob/master/docs-developer/loading-in-profiles.md#url).
I guess I can add `from-url`.

About `from-addon`: it works if you didn't do anything with the add-on since then. But if you capture a second profile (for example), or if you start/stop the addon again, you'll loose the data if you reload the page.

Reply:
I see, thanks.

> About `from-addon`: it works if you didn't do anything with the add-on since then. But if you capture a second profile (for example), or if you start/stop the addon again, you'll loose the data if you reload the page.

I thought the same at first too. But then I tried after closing the profiler and after capturing another profile and it seems still getting the same, correct profile after refreshing. But I'm not sure how it manages to do that. It might be cache or something. I think it's fine to exclude `from-addon` here.